build/v1/types.go

```diff
@@ -1335,3 +1348,59 @@ type SecretSpec struct {
 	// mountPath is the path at which to mount the secret
 	MountPath string `json:"mountPath" protobuf:"bytes,2,opt,name=mountPath"`
 }
+
+// BuildVolume describes a volume that is made available to build pods,
+// such that it can be mounted into buildah's runtime environment.
+// Only a subset of Kubernetes Volume sources are supported.
+type BuildVolume struct {
+	// name is a unique identifier for this BuildVolume.
+	// It must conform to the Kubernetes DNS label standard and be unique within the pod.
+	// It cannot collide with volumes that are always added by the build controller.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// source represents the location and type of the mounted volume.
+	Source BuildVolumeSource `json:"source" protobuf:"bytes,2,opt,name=source"`
+
+	// mounts represents the location and read/write status of the volume
+	// in the buildah container
+	Mounts []BuildVolumeMount `json:"mounts" protobuf:"bytes,3,opt,name=mounts"`
+}
+
+// BuildVolumeSourceType represents a build volume source type
+type BuildVolumeSourceType string
+
+const (
+	// BuildVolumeSourceTypeSecret is the Secret build source volume type
+	BuildVolumeSourceTypeSecret BuildVolumeSourceType = "Secret"
+
+	// BuildVolumeSourceTypeConfigmap is the ConfigMap build source volume type
+	BuildVolumeSourceTypeConfigMap BuildVolumeSourceType = "ConfigMap"
+)
+
+// BuildVolumeSource represents the source of a volume to mount
+// Only one of its supported types may be specified at any given time.
+type BuildVolumeSource struct {
+
+	// type is the BuildVolumeSourceType for the volume source.
+	// Type must match the populated volume source, and if not specified will be inferred on admission.
+	// +k8s:conversion-gen=false
+	// +optional
+	Type BuildVolumeSourceType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=BuildVolumeSourceType"`
+
+	// secret represents a Secret that should populate this volume.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#secret
+	// +optional
+	Secret *corev1.SecretVolumeSource `json:"secret,omitempty" protobuf:"bytes,2,opt,name=secret"`
+
+	// configMap represents a ConfigMap that should populate this volume
+	// +optional
+	ConfigMap *corev1.ConfigMapVolumeSource `json:"configMap,omitempty" protobuf:"bytes,3,opt,name=configMap"`
+}
+
+// BuildVolumeMount describes the mounting of a Volume within buildah's runtime environment.
+type BuildVolumeMount struct {
+	// destinationPath is the path within the buildah runtime environment at which the volume should be mounted.
+	// Must not contain ':' and can not collide with a destination path generated by the builder process.
+	DestinationPath string `json:"destinationPath" protobuf:"bytes,1,opt,name=destinationPath"`
+}
```

Comment:
A quick note regarding this API.

`Secrets` and `ConfigMaps` require read-only volumes to work in Kubernetes. Rather than forcing users to specify `ReadOnly`, we decided to drop this field and assume volumes will be mounted read-only by default.

If in the future we support writeable volumes (ex - PVCs), the API will need to expose a field like `Writeable` as an option, with validations to ensure that builds with Secret or ConfigMap volumes to not have this value set to true.

Reply:
so to be clear, you're mounting it as writable within the buildah container (so dockerfile commands can write to the mounted dir path) but those write operations will not propagate to the actual k8s resource backing the pod volume?

or will writes from the Dockerfile commands fail because even the buildah path is read-only?

i would guess it's the latter.

Regardless, the behavior should be specified.

Reply:
The volumeMounts within the containers pod definition are readOnly:
```
- mountPath: /var/run/openshift.io/volumes/my-secret-user-build-volume
      name: my-secret-user-build-volume
      readOnly: true
    - mountPath: /var/run/openshift.io/volumes/my-configmap-user-build-volume
      name: my-configmap-user-build-volume
      readOnly: true
```
And the transient mount definition that is sent to buildah is readonly also:
```
I0614 14:51:34.597118       1 daemonless.go:286] transient mounts: []string{"/tmp/rhsm-copy257734809:/run/secrets/rhsm:nodev,noexec,nosuid", "/var/run/openshift.io/volumes/my-secret-user-build-volume:/var/run/secrets/some-secret:ro", "/var/run/openshift.io/volumes/my-secret-user-build-volume:/var/run/secrets/some-secret-2:ro", "/var/run/openshift.io/volumes/my-configmap-user-build-volume:/var/run/configmaps/some-configmap:ro"}
```

Reply:
addressed in next push